syntax = "proto3";

package biblioteca.gRPC;

message Vazia {}

message Pesquisa {
    string termo = 1;
}

message Livro {
    string titulo = 1;
    string autor = 2;
    string isbn = 3;
}

message Usuario {
    string nome = 1;
    string email = 2;
    string cpf = 3;
}

message EmprestimoRequest {
    string cpf_usuario = 1;
    string titulo_livro = 2;
}

message BloqueioRequest {
    string cpf_usuario = 1;
}

message Resposta {
    string mensagem = 1;
}

service Biblioteca {
    rpc ListarLivros(Vazia) returns (stream Livro) {}
    rpc PesquisarLivros(Pesquisa) returns (stream Livro) {}
    rpc ListarUsuarios(Vazia) returns (stream Usuario) {}
    rpc PesquisarUsuarios(Pesquisa) returns (stream Usuario) {}

    // Funções de escrita
    rpc EmprestarLivro(EmprestimoRequest) returns (Resposta) {}
    rpc BloquearUsuario(BloqueioRequest) returns (Resposta) {}
}